### PR TITLE
Annotation for action target

### PR DESCRIPTION
**Pull request by yongpingchen**
#### Improve registration of Action/TargetState/ActionResult
In current design, registration of Action subclasses, ActionResult subclasses is not elegant.
The purpose of this PR is to make the registration simple and easy to use. Has the following improvement: 
1. Change`Action` to interface, which  is not used for only one single action. When developer define implementation of Action, must  defines the `ActionAnnotation` annotation. SDK use ActionAnnotation to serialize/deserialize the actions
2. `ActionResult` class is not abstract class. It is final class. SDK only use `ActionResult` in the command. Developer not need to defined subclasses of `ActionResult`. 
3. Change `TargetState` to interface. When Developer defines implementation class of `TargetState`, must define the `TargetStateAnnotation` annotation. SDK uses this annotation to serialize/deserialize the state.

#### Add wrapper of action/actionResult
- Add `AliasAction` to group Action by alias, which used in `Command`.
- Add `AliasActionResult` to group ActionResult by alias, which used in `Command`.

### Snippets
The following snippets show how it works.

Example action class and state class

```java
public class AirConditionerAction implements Action{
    @ActionAnnotation(actionName = "turnPower")
    public Boolean power;

    @ActionAnnotation(actionName = "setPresetTemperature")
    public Integer presetTemperature;
}

public class AirConditonerState implements TargetState {
    @TargetStateAnnotation()
    public boolean power;

    @TargetStateAnnotation()
    public int currentTemperature;
}

```

The snippet of registration actions/state:

```java
        Map<String, Class<? extends Action>> actionTypes = new HashMap<>();
        actionTypes.put("AirConditionerAlias", AirConditionerAction.class);

        Map<String, Class<? extends TargetState>> stateTypes = new HashMap<>();
        stateTypes.put("AirConditionerAlias", AirConditonerState.class);
        stateTypes.put("HumidityAlias", HumidityState.class);

        ThingIFAPI api = ThingIFAPIBuilder
                .newBuilder(
                        getApplicationContext(),
                        app,
                        new Owner(
                            new TypedID(TypedID.Types.USER, ownerID),
                            ownerToken),
                        actionTypes,
                        stateTypes
                        )
                .build();
```

Snippet to post command:

```java
        String alias = "AirConditionerAlias";

        AliasAction<AirConditionerAction> action1 = 
                new AliasAction<>(
                        alias, 
                        new AirConditionerAction(true, null)
                );

        AliasAction<AirConditionerAction> action3 = 
                new AliasAction<>(
                        alias, 
                        new AirConditionerAction(null, 25)
                );

        List<AliasAction<? extends Action>> actions = new ArrayList<>();
        actions.add(action1);
        actions.add(action3);
        CommandForm form = new CommandForm(actions);

        try {
            api.postNewCommand(form);
        }catch(Exception ex){
            //handle Exception
        }
```

Snippet to get command

```java
        String alias = "AirConditionerAlias";
        try {
            Command cmd = api.getCommand(commandID);
            
            // retrieve actions with alias
            List<AliasAction<AirConditionerAction>> actions = cmd.getAction(alias);
            for(AliasAction<AirConditionerAction> action : actions) {
                AirConditionerAction action1 = action.getAction();
                // handle the action
            }
            
            // iterate actions directly
            for(AliasAction aliasAction: cmd.getActions()) {
                if(aliasAction.getAlias().equals(alias)){
                    AirConditionerAction action1 = (AirConditionerAction) aliasAction.getAction();
                    // handle the action
                }
            }

            // iterate action results
            for(AliasActionResult results: cmd.getActionResults()) {
                if(results.getAlias().equals(alias)) {
                    for(ActionResult result: results.getResults()) {
                        if(result.getActionName().equals("turnPower")){
                            boolean succeeded = result.succeeded;
                            String errorMessage = result.getErrorMessage();
                        }
                    }
                }
            }
        }catch(Exception ex){
            //handle Exception
        }
```

Snippet to get target state
```java
            Map<String, ? extends TargetState> traitStates =
                    traitApi.getTargetState();

            for(Map.Entry<String, ? extends TargetState> e : traitStates.entrySet()){
                String alias = e.getKey();
                Object state = e.getValue();
                if(state instanceof AirConditonerState){
                    boolean power = ((AirConditonerState) state).power;
                    int temp = ((AirConditonerState) state).currentTemperature;
                }else if(state instanceof HumidityState){
                    int humidity = ((HumidityState) state).currentHumidity;
                }
            }

            AirConditonerState state = traitApi.getTargetState("AirConditionerAlias");
            boolean power = ((AirConditonerState) state).power;
            int temp = ((AirConditonerState) state).currentTemperature;
```

### CODE REVIEW

**Comment by yongpingchen** [comment]
@satoshikumano @shinsuke-nara please check this PR?

**Comment by yongpingchen** [comment]
@satoshikumano I fix the design of action/target state according to your comments. Please confirm

File: thingif/src/main/java/com/kii/thingif/trigger/TriggeredCommandForm.java [hunk 1]
```diff
@@ -43,14 +43,14 @@
      */
     public static class Builder {
 
-        @NonNull private List<Pair<String, List<Action>>> actions;
+        @NonNull private List<Pair<String, ? extends Action>> actions;
```
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/trigger/TriggeredCommandForm.java]: should be `AliasAction`

File: thingif/src/main/java/com/kii/thingif/trigger/TriggeredCommandForm.java [hunk 2]
```diff
@@ -71,7 +71,7 @@ private Builder(
          */
         @NonNull
         public static Builder newBuilder(
-                @NonNull List<Pair<String, List<Action>>> actions)
+                @NonNull List<Pair<String, ? extends Action>> actions)
```
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/trigger/TriggeredCommandForm.java]: override with non parameter

File: thingif/src/main/java/com/kii/thingif/trigger/TriggeredCommandForm.java [hunk 3]
```diff
@@ -122,7 +122,7 @@ public static Builder newBuilder(
          */
         @NonNull
         public Builder setActions(
```
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/trigger/TriggeredCommandForm.java]: change to `addActions`, which add single Action to List

File: thingif/src/main/java/com/kii/thingif/ThingIFAPI.java [hunk 4]
```diff
@@ -1402,7 +1399,7 @@ public String deleteTrigger(
      */
     @NonNull
     @WorkerThread
-    public TargetState getTargetState(
+    public <S extends TargetState> S getTargetState(
```
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/ThingIFAPI.java]: If S is not registered, throw `ClassCastException`. Mentioned it in API document.

File: thingif/src/main/java/com/kii/thingif/ThingIFAPIBuilder.java [hunk 5]
```diff
@@ -56,8 +53,7 @@ public static ThingIFAPIBuilder newBuilder(
             @NonNull Context context,
             @NonNull KiiApp app,
             @NonNull Owner owner,
-            @NonNull Map<String, Map<String,Class<? extends Action>>> actionTypes,
-            @NonNull Map<String, Map<String,Class<? extends ActionResult>>> actionResultTypes,
+            @NonNull Map<String, Class<? extends Action>> actionTypes,
             @NonNull Map<String, Class<? extends TargetState>> stateTypes) {
```
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/ThingIFAPIBuilder.java]: override newBuilder without atctionTypes, and actionResultTypes parameters

File: thingif/src/main/java/com/kii/thingif/ThingIFAPIBuilder.java [hunk 6]
```diff
@@ -173,28 +168,12 @@ public ThingIFAPI build() {
     @NonNull
     public ThingIFAPIBuilder registerActions(
             @NonNull String alias,
-            @NonNull Map<String,Class<? extends Action>> actionClasses){
+            @NonNull Class<? extends Action> actionClasses){
```
Comment (satoshikumano) [inline thingif/src/main/java/com/kii/thingif/ThingIFAPIBuilder.java]: actionClasses -> actionClass

File: thingif/src/main/java/com/kii/thingif/trigger/TriggeredCommandForm.java (hunk 1 above)
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/trigger/TriggeredCommandForm.java]: fixed on 6dfcb11f4cdef0451e4e717178c839826985df16

File: thingif/src/main/java/com/kii/thingif/trigger/TriggeredCommandForm.java (hunk 2 above)
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/trigger/TriggeredCommandForm.java]: fixed in commit 0bd6e7d484ecfdeb890e4f023f05a3e4a26140cf

File: thingif/src/main/java/com/kii/thingif/trigger/TriggeredCommandForm.java (hunk 3 above)
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/trigger/TriggeredCommandForm.java]: fixed in commit 0bd6e7d484ecfdeb890e4f023f05a3e4a26140cf

File: thingif/src/main/java/com/kii/thingif/ThingIFAPI.java (hunk 4 above)
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/ThingIFAPI.java]: fixed in commit 099bdc402d35974c52a0aa720c9ec7377c399b0d

File: thingif/src/main/java/com/kii/thingif/ThingIFAPIBuilder.java (hunk 6 above)
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/ThingIFAPIBuilder.java]: fixed in commit: 60eb33660d14ded374cabeab69993e0c91c36abe

File: thingif/src/main/java/com/kii/thingif/ThingIFAPIBuilder.java (hunk 5 above)
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/ThingIFAPIBuilder.java]: fixed in commit: 60eb33660d14ded374cabeab69993e0c91c36abe

**Comment by yongpingchen** [comment]
Investigate to use generic for `AliasAction<T extends Action>`

**Comment by yongpingchen** [comment]
@satoshikumano Can confirm this PR?

**Comment by satoshikumano** [comment]
ThingIFAPIBuilder#addSchema can be removed.

**Comment by yongpingchen** [comment]
@satoshikumano Not only that method, the Schema and SchemaBuilder class also needed to be removed. I will remove them in separate PR, since it will affect a lot of files.

**Comment by satoshikumano** [comment]
CommandForm.Builder would be useful to add action one by one in fluent style.

**Comment by satoshikumano** [comment]
There's two patterns of providing Builder class.

* With inner class: ClassToBeBuilt.Builder
* With separated class: ClassToBeBuiltBuilder

Can we unify the way of providing Builder class?

**Comment by yongpingchen** [comment]
The inner class pattern is good, which not increasing the number of file.

**Comment by satoshikumano** [comment]
In iOS, we have decided to merge Count and Aggregation.
KiiPlatform/thing-if-iOSSDK/#265

- Field has been unified.
- ThingIFAPI#count has been removed.

Can you apply this to Android as well?

**Comment by yongpingchen** [comment]
> In iOS, we have decided to merge Count and Aggregation.
KiiPlatform/thing-if-iOSSDK/#265
>  - Field has been unified.
>  - ThingIFAPI#count has been removed.
> Can you apply this to Android as well?

@satoshikumano I will do it another PR.

**Comment by yongpingchen** [comment]
@satoshikumano I pushed a new commit to add Builder to CommandForm. Can you check?

**Pull request merged**